engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -61,70 +70,90 @@ public Table table() {
 
     @Override
     public Collection<String> keyColumns() {
-        return null;
+        return keyColumns;
     }
 
     private MultiJoinTableImpl(@NotNull final JoinControl joinControl,
             @NotNull final MultiJoinInput... multiJoinInputs) {
+        keyColumns = new ArrayList<>();
+        if (multiJoinInputs[0].columnsToMatch().length == 0) {
+            table = doMultiJoinZeroKey(multiJoinInputs);
+            return;
+        }
         table = multiJoin(joinControl, multiJoinInputs);
     }
 
     private Table multiJoin(@NotNull final JoinControl joinControl,
             @NotNull final MultiJoinInput... multiJoinInputs) {
         final TObjectIntHashMap<String> usedColumns =
                 new TObjectIntHashMap<>(multiJoinInputs[0].columnsToAdd().length, 0.5f, -1);
+
+        final ColumnSource<?>[] firstKeySources = getKeySources(multiJoinInputs[0]);
+        final ColumnSource<?>[] originalColumns = getOriginalKeyColumns(multiJoinInputs[0]);
+
+        // Gather and verify the key columns
         final List<String> expectedLeftMatches =
                 ColumnName.names(JoinMatch.lefts(Arrays.asList(multiJoinInputs[0].columnsToMatch())));
-        if (expectedLeftMatches.size() == 0) {
-            return doMultiJoinZeroKey(multiJoinInputs);
-        }
-
-        final Set<String> keyColumnNames = new LinkedHashSet<>(expectedLeftMatches);
+        keyColumns.addAll(expectedLeftMatches);
         for (final String keyColumn : expectedLeftMatches) {
             usedColumns.put(keyColumn, KEY_COLUMN_SENTINEL);
         }
-        final ChunkType[] expectedChunkTypes = Arrays.stream(getKeySources(multiJoinInputs[0]))
-                .map(ColumnSource::getChunkType).toArray(ChunkType[]::new);
 
-        final MultiJoinInput[] useDescriptors = Arrays.copyOf(multiJoinInputs, multiJoinInputs.length);
-
-        for (int jj = 0; jj < multiJoinInputs.length; ++jj) {
-            MultiJoinInput joinDescriptor = useDescriptors[jj];
+        final Class<?>[] expectedColumnTypes = Arrays.stream(firstKeySources)
+                .map(ColumnSource::getType).toArray(Class<?>[]::new);
+        final Class<?>[] expectedComponentTypes = Arrays.stream(firstKeySources)
+                .map(ColumnSource::getComponentType).toArray(Class<?>[]::new);
 
+        for (int ii = 1; ii < multiJoinInputs.length; ++ii) {
+            ColumnSource[] currentKeySources = getKeySources(multiJoinInputs[ii]);
+            // Verify matching column names.
             final List<String> currentLeftMatches =
-                    ColumnName.names(JoinMatch.lefts(Arrays.asList(joinDescriptor.columnsToMatch())));
+                    ColumnName.names(JoinMatch.lefts(Arrays.asList(multiJoinInputs[ii].columnsToMatch())));
```

Comment:
Can we allow differing orders? Since we have a LHS name, we *could* allow for out-of-order matches.